engine/runtime/src/main/scala/org/enso/compiler/Cache.scala

```diff
@@ -0,0 +1,463 @@
+          logLevel,
+          s"Could not load the cache metadata " +
+          s"at [${metadataPath.toMaskedPath.applyMasking()}]"
+        )
+        invalidateCache(cacheRoot)
+        None
+    }
+  }
+
+  /** Validates the deserialized data by returning the expected cached entry, or [[Failure]].
+    *
+    * @param obj deserialized object
+    * @param meta metadata corresponding to the `obj`
+    * @return an `obj` transformed to a cached entry, or [[Failure]]
+    */
+  protected def validateReadObject(obj: AnyRef, meta: M): Try[T]
+
+  /** Read metadata representation from the provided location
+    * @param path location of the serialized metadata
+    * @return deserialized metadata, or [[None]] if invalid
+    */
+  private def loadCacheMetadata(
+    path: TruffleFile
+  ): Option[M] = {
+    if (path.isReadable) {
+      val bytes = path.readAllBytes()
+      metadataFromBytes(bytes)
+    } else {
+      None
+    }
+  }
+
+  protected def metadataFromBytes(bytes: Array[Byte]): Option[M]
+
+  private def ensureRoot(cacheRoot: TruffleFile): Boolean = {
+    try {
+      if (cacheRoot.exists() && cacheRoot.isDirectory()) {
+        return cacheRoot.isWritable
+      } else {
+        cacheRoot.createDirectories()
+        return cacheRoot.isWritable
+      }
+    } catch {
+      case _: IOException | _: UnsupportedOperationException |
+          _: SecurityException | _: FileAlreadyExistsException =>
+    }
+    false
+  }
+
+  /** Computes SHA-224 digest from the provided bytes.
+    */
+  protected def computeDigestFromBytes(bytes: Array[Byte]): String = {
+    Hex.toHexString(messageDigest().digest(bytes))
+  }
+
+  /** Raw (byte)  representation of the [[Metadata]].
+    *
+    * @param sourceDigest SHA-256 digest of source data
+    * @param blobDigest SHA-256 digest of serialized data
+    * @param entry cache entry
+    * @return serialized metadata
+    */
+  protected def metadata(
+    sourceDigest: String,
+    blobDigest: String,
+    entry: T
+  ): Array[Byte]
+
+  /** Gets the path to the cache data within the `cacheRoot`.
+    *
+    * @param cacheRoot the root of the cache for this entry
+    * @return the name of the data file for this entry's cache
+    */
+  private def getCacheDataPath(cacheRoot: TruffleFile): TruffleFile = {
+    cacheRoot.resolve(cacheFileName(dataSuffix))
+  }
+
+  private def getCacheMetadataPath(cacheRoot: TruffleFile): TruffleFile =
+    cacheRoot.resolve(cacheFileName(metadataSuffix))
+
+  /** Computes the cache file name for a given extension.
+    *
+    * @param suffix the extension
+    * @return the cache file name with the provided `ext`
+    */
+  private def cacheFileName(suffix: String): String = {
+    s"$entryName$suffix"
+  }
+
+  protected def stringRepr: String
+
+  protected def entryName: String
+
+  protected def dataSuffix: String
+
+  protected def metadataSuffix: String
+
+  protected def messageDigest(): MessageDigest = {
+    new SHA3.Digest224()
```

Comment:
We both noticed that computing hash occupies significant amount of time during startup. Wouldn't it be simpler and faster to compute MD5? Why do we need anything more complex? We are not concerned about malicious injection of wrong caches, right? Only about accidental corruption of the data. Then anything (even XOR, CRC, etc.) is good enough, I think.

Reply:
I've just found the [discussion](https://github.com/enso-org/enso/blob/1bc27501e6c8a6018e4c17e68ef4cb4068db15f9/distribution/lib/Standard/Table/0.0.0-dev/src/Data/Table.enso#L118).